### PR TITLE
Fuzz merge 20240811

### PR DESCRIPTION
**Pull request by pauldreik**
This adds a fuzzer for the conversion family of functions. It also invokes validate, count as needed.
It is this fuzzer I have used to find the buffer overruns, undefined behaviour and implementation differences recently.
It replaces the validation fuzzer.
The fuzzers are added to (optionally) build as part of the project, to avoid code rot and to simplify developing the fuzzer in an IDE.
The main way to use the fuzzers is still through the build.sh script since that sets the appropriate flags and keeps compatibility with oss-fuzz.

A separate main function is provided, with the intention to later add a CI job which runs old fuzz  data through the system also on platforms where it is inconvenient to run fuzzing (anything besides linux on amd64).

### CODE REVIEW

**Comment by lemire** [comment]
@pauldreik There is a crazy CIFuzz failure on this PR. Do you know what it is about?

**Comment by pauldreik** [comment]
> @pauldreik There is a crazy CIFuzz failure on this PR. Do you know what it is about?

yes! it is problems found by the fuzzer. I have reported them already. See https://github.com/simdutf/simdutf/issues/493  https://github.com/simdutf/simdutf/pull/494 https://github.com/simdutf/simdutf/pull/492

**Comment by pauldreik** [comment]
so, those need to be fixed before this is merged. otherwise we get issues reported by oss-fuzz.

File: fuzz/conversion.cpp [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+  Conversion<ENCODING_FROM_CONVERSION_NAME(&I::x),                             \
+             ENCODING_TO_CONVERSION_NAME(&I::x), decltype(&I::len),            \
+             decltype(&I::x)> {                                                \
+    nullptr, nullptr, std::string{NAMEOF(&I::len)}, std::string {              \
+      NAMEOF(&I::x)                                                            \
+    }                                                                          \
+  }
+  std::tuple cases_requiring_valid_data(
+      // all these cases require valid input for invoking the convert function
+      ADD(latin1_length_from_utf16, convert_valid_utf16be_to_latin1),
```
Comment (lemire) [inline fuzz/conversion.cpp:567]: `convert_valid_utf16be_to_latin1` is not part of our public API (we don't document it in the README) and it requires that the input is valid and can be converted to latin1.

File: fuzz/conversion.cpp [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+      ADD(utf32_length_from_utf16be, convert_valid_utf16be_to_utf32),
+      ADD(utf8_length_from_utf16be, convert_valid_utf16be_to_utf8),
+
+      ADD(latin1_length_from_utf16, convert_valid_utf16le_to_latin1),
```
Comment (lemire) [inline fuzz/conversion.cpp:571]: `convert_valid_utf16be_to_latin1` is not part of our public API (we don't document it in the README) and it requires that the input is valid and can be converted to latin1.

File: fuzz/conversion.cpp [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+      ADD(utf32_length_from_utf16le, convert_valid_utf16le_to_utf32),
+      ADD(utf8_length_from_utf16le, convert_valid_utf16le_to_utf8),
+
+      ADD(latin1_length_from_utf32, convert_valid_utf32_to_latin1),
```
Comment (lemire) [inline fuzz/conversion.cpp:575]: `convert_valid_utf32_to_latin1` is not part of our public API (we don't document it in the README) and it requires that the input is valid and can be converted to latin1.

File: fuzz/conversion.cpp [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+      ADD(utf16_length_from_utf32, convert_valid_utf32_to_utf16be),
+      ADD(utf16_length_from_utf32, convert_valid_utf32_to_utf16le),
+      ADD(utf8_length_from_utf32, convert_valid_utf32_to_utf8),
+
+      IGNORE(latin1_length_from_utf8, convert_valid_utf8_to_latin1),
```
Comment (lemire) [inline fuzz/conversion.cpp:580]: `convert_valid_utf8_to_latin1` is not part of our public API (we don't document it in the README) and it requires that the input is valid and can be converted to latin1.

**Comment by lemire** [comment]
Please see https://github.com/simdutf/simdutf/pull/496

It is fine to fuzz the convert valid to latin 1 functions, but only if the input can be converted to latin1.

I am deprecated them as I do not think that they are generally useful.

File: fuzz/conversion.cpp (hunk 1 above)
Comment (pauldreik) [reply fuzz/conversion.cpp]: disabled

File: fuzz/conversion.cpp (hunk 3 above)
Comment (pauldreik) [reply fuzz/conversion.cpp]: disabled

File: fuzz/conversion.cpp (hunk 4 above)
Comment (pauldreik) [reply fuzz/conversion.cpp]: disabled

File: fuzz/conversion.cpp [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+            if (!outputs_agree && !allow_implementations_to_differ)
+                break;
+
+            // coming this far means no problems were found
+            return;
+        } while (0);
+        // if we come here, something failed
+        std::cerr
+            << "something failed, rerun with PRINT_FUZZ_CASE set to print a reproducer to stderr\n";
+        std::abort();
```
[the remark this answers is not in the document]
Comment (pauldreik) [reply fuzz/conversion.cpp:246]: this is intentional, abort both works for unattended fuzzing as well as during debugging. assert may be disabled.

**Comment by pauldreik** [comment]
This should now be ready for a second round of review. On amd64 linux, it now runs without finding anything on the latest master.

**Comment by lemire** [comment]
Fantastic work.

**Pull request merged**